src/main/java/io/anserini/rerank/lib/TiebreakerReranker.java

```diff
@@ -0,0 +1,131 @@
+ * then truncate to top <i>k</i>.</p>
+ */
+public class TiebreakerReranker implements Reranker {
+    // Sort by score, break ties by lexicographic ordering of collection docid.
+    private class Result implements Comparable<Result> {
+      public float score;
+      public String docid;
+      public int ldocid;
+      public Document doc;
+
+      public Result(float score, int ldocid, String docid, Document doc) {
+        this.score = score;
+        this.ldocid = ldocid;
+        this.docid = docid;
+        this.doc = doc;
+      }
+
+      public int compareTo(Result other) {
+        if (this.score > other.score) {
+          return -1;
+        } else if (this.score < other.score) {
+          return 1;
+        }
+
+        return this.docid.compareTo(other.docid);
+      }
+    }
+
+    @Override
+    public ScoredDocuments rerank(ScoredDocuments docs, RerankerContext context) {
+      SortedSet<Result> sortedResults = new TreeSet<>();
+      for (int i=0; i<docs.documents.length; i++ ) {
+        float rounded = Math.round(docs.scores[i] * 1e4f) / 1e4f;
+        sortedResults.add(new Result(rounded, docs.ids[i],
+            docs.documents[i].getField(FIELD_ID).stringValue(), docs.documents[i]));
+      }
+
+      int numResults = sortedResults.size();
+      ScoredDocuments rerankedDocs = new ScoredDocuments();
+      rerankedDocs.documents = new Document[numResults];
+      rerankedDocs.ids = new int[numResults];
+      rerankedDocs.scores = new float[numResults];
+
+      int i = 0;
+      int dup = 0;
+      float prevScore = 0.0f;
+      for (Result result : sortedResults) {
+        float curScore = result.score;
+
+        // If we encounter ties, we want to perturb the final score a tiny bit.
+        // Here's the basic approach, by example. Let's say our starting ranked list was:
+        //
+        //   1 docA 23.439316
+        //   2 docS 22.087432
+        //   3 docT 22.087432
+        //   4 docZ 21.602508
+        //
+        // The point is that we want to perturb the scores in a small way such that the scores give us the exact sort
+        // order we want, independent of how any external evaluation tool (e.g., trec_eval) breaks ties.
+        // We accomplish this by rounding all scores to 1e-4, and then subtracting a minor delta of 1e-6 for each tie.
+        // So, the above becomes:
+        //
+        //   1 docA 23.4393   (dup=0)
+        //   2 docS 22.0874   (dup=0)
+        //   3 docT 22.0874 - (dup=1)*1e-6
+        //   4 docZ 21.6025   (dup=0)
+        //
+        // Note that we can't use equality comparison directly to detect duplicates, because in the case of multiple
+        // ties, we would have perturbed the scores, leading the scores to not be equal (hence we check for score
+        // difference greater than 1e-4).
+        //
+        // Why 1e-4 and 1e-6? If we make the former larger, than we lose score resolution in the original score. If we
```

Comment:
What if there are more than 100 documents with the same score 22.0874, and the document below has score 22.0873?

Reply:
Correct... this trick can't handle more than 100 score ties... unless we do 1e-3 and 1e-6, in which case we can support up to 1000 scoring ties..

Reply:
Honestly, I do not think the manipulation here is necessary....and it is hard to maintain and debug.
`Result` class can handle most cases and the only problem with it is that it cannot handle score ties across the retrieval hits.
Now you doubled the first round retrieval results and `Result` should be able to solve the problem mentioned above.
The only problem for now is how safe is the number of hits in the first round - I don't know and nobody knows... But 1000 should be a pretty safe bet as for MB collections I've seen more than 100 score ties but I've never seen more than several hundreds score ties.